### PR TITLE
Write Event Lifecycle Test +1 Endpoint

### LINKED ISSUE #110797: Write Event Lifecycle Test +1 Endpoint
# Progress <code>[6/6]</code>

-   [X] APISnoop org-flow : [replaceCoreV1NamespacedEventTest.org](https://github.com/apisnoop/ticket-writing/blob/master/replaceCoreV1NamespacedEventTest.org)
-   [X] test approval issue : #110797 
-   [X] test pr : #110798
-   [X] two weeks soak start date : 1 Jul 2022 [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&graph-metrics=test-duration-minutes&include-filter-by-regex=should.manage.the.lifecycle.of.an.event)
-   [x] two weeks soak end date : 15 Jul 2022
-   [x] test promotion pr : #110797 

# Identifying an untested feature Using APISnoop

According to this APIsnoop query, there are still a single remaining Event endpoint which is untested.

```sql-mode
  select
    endpoint,
    path,
    kind
  from testing.untested_stable_endpoint
  where eligible is true
  and endpoint ilike '%Event%'
  order by kind, endpoint desc
  limit 10;
```

```example
             endpoint           |                     path                     | kind
  ------------------------------+----------------------------------------------+-------
   replaceCoreV1NamespacedEvent | /api/v1/namespaces/{namespace}/events/{name} | Event
  (1 row)

```

# API Reference and feature documentation

-   [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
-   [Kubernetes API > Cluster Resources > Event](https://kubernetes.io/docs/reference/kubernetes-api/cluster-resources/event-v1/)
-   [client-go: core/v1/event.go](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/event.go#L42-L54)

# Test outline

```
Feature: Test replace Event api endpoint
```

-   replaceCoreV1NamespacedEvent

```
Scenario: the test updates an event
  Given the e2e test has an event
  And a new event series has been created
  When the test updates the event
  Then the requested action is accepted without any error
  And a newly fetched copy passes and equality check with the current event
```

# E2E Test

Using a number of existing e2e test practices a new [ginkgo test](https://github.com/ii/kubernetes/blob/create-event-lifecycle-test/test/e2e/instrumentation/core_events.go#L135-L242) has been created for this pod status endpoint. The e2e logs for this test are listed below.

```
[It] should manage the lifecycle of an event
  /home/ii/go/src/k8s.io/kubernetes/test/e2e/instrumentation/core_events.go:135
STEP: creating a test event
STEP: listing all events in all namespaces
STEP: patching the test event
STEP: fetching the test event
STEP: updating the test event
STEP: getting the test event
STEP: deleting the test event
STEP: listing all events in all namespaces
```

# Verifying increase in coverage with APISnoop

## Listing endpoints hit by the updated e2e test

This query shows the event endpoints hit within a short period of running the e2e test.

```sql-mode
select distinct  endpoint, right(useragent,39) AS useragent
from testing.audit_event
where endpoint ilike '%Event%'
and release_date::BIGINT > round(((EXTRACT(EPOCH FROM NOW()))::numeric)*1000,0) - 60000
and useragent like 'e2e%should%'
order by endpoint
limit 10;
```

```example
            endpoint             |                useragent
---------------------------------+-----------------------------------------
 createCoreV1NamespacedEvent     | should manage the lifecycle of an event
 deleteCoreV1NamespacedEvent     | should manage the lifecycle of an event
 listCoreV1EventForAllNamespaces | should manage the lifecycle of an event
 patchCoreV1NamespacedEvent      | should manage the lifecycle of an event
 readCoreV1NamespacedEvent       | should manage the lifecycle of an event
 replaceCoreV1NamespacedEvent    | should manage the lifecycle of an event
(6 rows)

```

# Final notes

If a test with these calls gets merged, test coverage will go up by 1 point.

This test is also created with the goal of conformance promotion.

---

/sig testing
/sig architecture
/area conformance

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- replaceCoreV1NamespacedEvent

**Which issue(s) this PR fixes:**
Fixes #110797

**Testgrid Link:** 
[testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&graph-metrics=test-duration-minutes&include-filter-by-regex=should.manage.the.lifecycle.of.an.event)

**Special notes for your reviewer:**
Adds +1 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

### CODE REVIEW

**Comment by Riaankl** [comment]
/hold

This PR is an example of the proposed method for updating conformance tests as discussed in the SIG architecture meeting of 16 June 2022.
The endpoint `replaceCoreV1NamespacedEvent` was missing a conformance test. There is, however , an existing conformance test for covering 5 other `Event` endpoints.

In order to uphold the requirements that conformance test must have a 2 week flaker free run as an e2e test, before being promoted to conformance, the original Conformance test was copied, downgraded to an e2e test and the section added for `replaceCoreV1NamespacedEvent`.

Once this PR is merged there would be duplicate testing of `Event` endpoints for the next 2 weeks. If the test passes the two week flake free requirements there are two options to proceed:

1. Update the e2e test to conformance and remove the then obsolete conformance test that is missing the `replaceCoreV1NamespacedEvent` endpoint in one PR. 

This will keep the context together in a single PR. But it would require an update in Conformance policy, as the policy states that a conformance promotion must:
```contains no other modifications to test source code other than the following:
- modifies the testcase to use the framework.ConformanceIt() function rather than the framework.It() function
- adds a comment immediately before the ConformanceIt() call that includes all of the required [conformance test comment metadata](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md#conformance-test-comment-metadata)
- run hack/update-conformance-yaml.sh which adds the test name to the [conformance.yaml](https://github.com/kubernetes/kubernetes/blob/master/test/conformance/testdata/conformance.yaml) file
```

Or

2. Promote the e2e test to conformance after 2 week flake-free success in one PR and remove the obsolete test in a second PR. This would not require a change of policy, but the history would not be as clear later on.

This is a larger issue than just this one PR. There are many resources that have several endpoints tested in separate e2e tests, resulting in the repetition of create & delete operations for every test.

Once conformance text debt is paid off this work could be undertaken. 
It would help reduce infrastructure workload by consolidating conformance tests in the manner proposed here. 

/assign @dims @johnbelamaric @smarterclayton

**Comment by dgrisonnet** [comment]
@Riaankl leaving it up to you to unhold this PR once the update path has been figured out.

/approve

**Review by logicalhan — APPROVED** [review]
/lgtm
/approve

**Comment by Riaankl** [comment]
@johnbelamaric @dims @smarterclayton 
We discussed a method of updating of conformance tests by coping the test, change it back to an e2e test & update.
Once the e2e test provened flakefree on the test grid, we will promote it with the change. In the last SIG Arch meeitng we agreed the method is a good approach. Therefor I believe it is OK to merge this PR.

The only open question is if we should update the policy to allow for the removal of old test in the same PR as the Promotion PR of if the removal must be in a second PR. 
IMO If we use the same PR of promotion and removal it give better historical tracking.

/unhold

**Comment by Riaankl** [comment]
Unrelated flake
```
=== SKIP: vendor/k8s.io/legacy-cloud-providers/vsphere TestVolumes (0.00s)
    vsphere_test.go:941: No config found in environment
=== Failed
=== FAIL: vendor/k8s.io/kubectl/pkg/cmd/wait TestWaitForJSONPathCondition/handles_watch_condition_change (0.10s)
    wait_test.go:1683: theresource.group "foo-b6699dcfb-rnv7t" not found
    --- FAIL: TestWaitForJSONPathCondition/handles_watch_condition_change (0.10s)
=== FAIL: vendor/k8s.io/kubectl/pkg/cmd/wait TestWaitForJSONPathCondition (3.57s)
DONE 34957 tests, 19 skipped, 2 failures in 1.942s
+++ [0630 19:07:41] prune-junit-xml not found; installing from hack/tools
processing junit xml file : /logs/artifacts/junit_20220630-184011.xml
done.
+++ [0630 19:07:42] Saved JUnit XML test report to /logs/artifacts/junit_20220630-184011.xml
make: *** [Makefile:186: test] Error 1
```
/test pull-kubernetes-unit

**Comment by Riaankl** [comment]
Unreleated flake
```
=== SKIP: vendor/k8s.io/legacy-cloud-providers/vsphere TestVolumes (0.00s)
    vsphere_test.go:941: No config found in environment
=== Failed
=== FAIL: pkg/kubelet/nodeshutdown TestFeatureEnabled/shutdownGracePeriodRequested_0;_disables_feature (0.00s)
==================
WARNING: DATA RACE
```
/test pull-kubernetes-unit

**Pull request merged**
